### PR TITLE
Fix #2425. Load flash dynamically.

### LINKED ISSUE #2425: flashcap dynamically load FLASH into job_cmd container
When a user has flash privileges, we will have to convey some credentials to the job_agent, which will get to job_cmd., which will have to pull in FLASH into the container if not already there. 

Here's a rough outline:

* User has to have flash role
* FLASH tarball/rpm will be linked/copied into the user's flash/lib by the server
* sim_data (server) will add this library file to the lib_files list
* prepare_simulation (job_cmd) will pull the file and unpack it if FLASH is not already available

### PR DESCRIPTION
**Pull request by e-carlin**
This branch is base on issue/2361-prune-non-premium for the `auth.set_user` function. Please review and merge #2433 before reviewing.

The CapLaser example fails due to a parameter misconfiguration. I didn't take the time to troubleshoot because it wasn't relevant to this PR which was just for architecture. I haven't tried RTFlame.

### CODE REVIEW

File: sirepo/package_data/template/flash/lib/CapLaser.setup_units [hunk 1]
```diff
@@ -0,0 +1,161 @@
+Driver
```
Comment (e-carlin) [inline sirepo/package_data/template/flash/lib/CapLaser.setup_units:1]: @robnagler do we want this file here? We parse it [when generating parameters.py for export](https://github.com/radiasoft/sirepo/blob/47a24d8bd21f8916dddad2146a0106544b5924bd/sirepo/template/flash.py#L427). Maybe it should be packaged with the binary and distributed with it?

File: sirepo/pkcli/admin.py [hunk 2]
```diff
@@ -22,6 +23,51 @@
...
+
+    def _link_or_unlik_protected_files(sim_type, should_link):
+        for f in sim_data.get_class(sim_type).protected_lib_file_basenames():
+            p = simulation_db.simulation_lib_dir(sim_type).join(f)
+            if not should_link:
+                pkio.unchecked_remove(p)
+                continue
+            try:
+                p.mksymlinkto(
+                    # TODO(e-carlin): discuss with rn file structure for flash binaries
```
Comment (e-carlin) [inline sirepo/pkcli/admin.py:44]: @robnagler what do we want the file structure to look like in dev, alpha, etc?

File: sirepo/sim_data/__init__.py [hunk 3]
```diff
@@ -371,7 +371,12 @@ def lib_files_to_run_dir(cls, data, run_dir):
             data (dict): simulation db
             run_dir (py.path): where to copy to
         """
-        for b in cls.lib_file_basenames(data):
+        f = cls.lib_file_basenames(data)
+        # TODO(e-carlin): When remote we will pull the flash binary
```
Comment (e-carlin) [inline sirepo/sim_data/__init__.py:375]: @robnagler is this a problem?

File: sirepo/pkcli/admin.py [hunk 4]
```diff
@@ -22,6 +23,51 @@
...
+                    absolute=False,
+                )
+            except py.error.EEXIST:
+                pass
+
+
+    def _audit_user(uid):
+        with auth.set_user(uid):
+            for t in feature_config.cfg().proprietary_sim_types:
+                _link_or_unlik_protected_files(
```
Comment (robnagler) [inline sirepo/pkcli/admin.py:55]: typo "unlink"

File: sirepo/pkcli/admin.py [hunk 5]
```diff
@@ -22,6 +23,51 @@
...
+                        auth.role_for_sim_type(t),
+                        raise_forbidden=False,
+                    ),
+                )
+
+    server.init()
+    if uid:
+        _audit_user(uid)
+        return
+    for d in pkio.sorted_glob(simulation_db.user_dir_name().join('*')):
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/pkcli/admin.py:67]: Do we want the abstraction to be from simulation_db (glob of the uids in the db) or from auth_db (query sqlite for all uids)?

[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/pkcli/admin.py:67]: I think we should rely auth_db more now that you've removed the flask coupling. Make sense?

File: sirepo/sim_data/flash.py [hunk 6]
```diff
@@ -20,6 +20,16 @@ def fixup_old_data(cls, data):
                 plot_var_6='depo',
             )
 
+    @classmethod
+    def protected_lib_file_basenames(cls):
+        return ['CapLaser']
+
+    @classmethod
+    def setup_units_file_path(cls, flash_type):
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/sim_data/flash.py:28]: flash_type is used elsewhere (ex [here](https://github.com/radiasoft/sirepo/blob/47a24d8bd21f8916dddad2146a0106544b5924bd/sirepo/package_data/static/html/flash-physics.html#L3) and [here](https://github.com/radiasoft/sirepo/blob/47a24d8bd21f8916dddad2146a0106544b5924bd/sirepo/template/flash.py#L282)). I think I'm using it to mean the same thing here. Maybe we should remove the concept of flashType altogether and always use example name?  Or am I missing something and this is a different use?

flash_type has a one-to-one relationship with example name so maybe it is better to always just use example name.

[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/sim_data/flash.py:28]: Or are you referring to just passing in data as you mention in [this comment](https://github.com/radiasoft/sirepo/pull/2461#discussion_r416228089)?

File: sirepo/pkcli/admin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/pkcli/admin.py]: Implemented in https://github.com/radiasoft/sirepo/pull/2461/commits/2d6f471641e3655cbb69ae6d0154f43c5d889cc6. I had to do some rework of [user_dir_not_found to raise](https://github.com/radiasoft/sirepo/pull/2461/files#diff-f3f7b7715c3b54e0a71de6e4ecda4758R434) an [exception we could catch](https://github.com/radiasoft/sirepo/pull/2461/files#diff-05907a97c178a66d79820a99beb80abeR41) in the case where a uid exists in auth_db but not in sim_db.

File: sirepo/pkcli/admin.py [hunk 7]
```diff
@@ -83,3 +129,12 @@ def _create_example(example):
 
 def _is_src_dir(d):
     return re.search(r'/src$', str(d))
+
+
+cfg = pkconfig.init(
+    flash_src_dir=(
+        '/home/vagrant/src/FLASH4.6.2/',
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/pkcli/admin.py:140]: Having the example name in the path is tricky. The simulation lib dir (and resource dir) is a flat directory (no example name in path) so we will need to have unique filenames across example dirs, do some munging with the example name and file name, or make the lib dir structure no longer flat. Do you have a preference?

We do need a "setup_dev". Let's chat about how to do it (ex authentication, where to keep the files, etc).

File: sirepo/pkcli/admin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/pkcli/admin.py]: > in the case where a uid exists in auth_db but not in sim_db.

I don't think we should catch this specially. In general for these "all uid" things, we will want to report errors. If there is a broken uid (no dir in db), then it should stop.

**Pull request closed without merging**